src/ngamsUtils/ngamsUtils/ngasXSyncTool.py

```diff
@@ -1101,24 +1113,32 @@ def check_if_file_in_target_cluster(connection, cluster_nodes, file_id, file_ver
     :return: Information for file as tuple:
                 (<Disk Id>, <File ID>, <File Version>) or None if the file is not available in the cluster (tuple|None)
     """
-    logger.info("Entering check_if_file_in_target_cluster() ...")
+    logger.debug("Entering check_if_file_in_target_cluster() ...")
 
     # Reformat the cluster node list to a string for passing into the SQL statement
-    # It should be in the format "'ngas04:7777', 'ngas04:7778'"
+    # It should be in the format "'ngas04:7777','ngas04:7778'"
     cluster_nodes_sql = str(cluster_nodes.split(',')).replace('[', '').replace(']', '')
 
+    # FIXME: this fails to format the SQL query probably because we are inserting a list of nodes
+    # sql = "select nf.disk_id, nf.file_id, nf.file_version " \
+    #       "from ngas_files nf, ngas_disks nd " \
+    #       "where nf.file_id = {0} and nf.file_version = {1} and nf.disk_id = nd.disk_id " \
+    #       "and (nd.host_id in ({2}) or nd.last_host_id in ({3}))"
+    # result = connection.query2(sql, args=(file_id, file_version, cluster_nodes_sql, cluster_nodes_sql))
+
     sql = "select nf.disk_id, nf.file_id, nf.file_version " \
           "from ngas_files nf, ngas_disks nd " \
-          "where nf.file_id = '{0}' and nf.file_version = {1} and nf.disk_id = nd.disk_id " \
-          "and (nd.host_id in ({2}) or nd.last_host_id in ({3}))"
-    result = connection.query2(sql, args=(file_id, file_version, cluster_nodes_sql, cluster_nodes_sql))
+          "where nf.file_id = '{:s}' and nf.file_version = {:d} and nf.disk_id = nd.disk_id " \
+          "and (nd.host_id in ({:s}) or nd.last_host_id in ({:s}))"
+    sql = sql.format(file_id, file_version, cluster_nodes_sql, cluster_nodes_sql)
```

Comment:
If this works for you, that's great, but I have to ask: what error did you initially get? Maybe the way the parameters were given in the previous version were wrong? I'd imagine that for parameter binding to work `cluster_nodes_sql` would have had to be a list of strings instead of a manually formatted subset of the SQL query (or in other words I would have expected you to be able to use `cluster_nodes` directly in `query2`, and maybe even the parentheses around `{2}` and `{3}` might have been not necessary). Would you mind providing a few more details around that?

Reply:
I try passing a list as an argument into a formatted string with and without parantheses. This does not work for me. You can seem my attempts in the commented code. I get the error messages `DatabaseError: ORA-01484: arrays can only be bound to PL/SQL statements`. I am unsure what else I can do.

Reply:
Mmmm.... after some more reading, apparently the way to go is to have a placeholder for each value in the list, then pass down all individual values, like so: https://cx-oracle.readthedocs.io/en/latest/user_guide/bind.html#binding-multiple-values-to-a-sql-where-in-clause (in specific check where it says "If the number of values is only going to be known at runtime"). Would you mind trying that approach? Beware that in the case of NGAS we use `{0}`, `{1}` and so on for the placeholders, which we then internally translate into the driver's specific nomenclature.

Reply:
I think I have resolved this. The code is much bulkier and not so easy to follow. Anyway, at least it now works

Reply:
Thanks @smclay. Yes, the code is a bit more complex to follow, but at least now it should be free from SQL injection bugs, which is the whole point of using prepared statements. Thanks for your patience!